x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/LastValueStates.java

```diff
@@ -0,0 +1,195 @@
+import java.io.IOException;
+import java.util.Objects;
+
+public final class LastValueStates {
+    private LastValueStates() {}
+
+    static class SingleState implements AggregatorState {
+        private double lastValue = Double.NaN;
+
+        private long lastTimestamp = 0;
+
+        SingleState() {}
+
+        private SingleState(BytesRef other) {
+            try {
+                try (ByteArrayStreamInput in = new ByteArrayStreamInput(other.bytes)) {
+                    in.reset(other.bytes, other.offset, other.length);
+                    lastValue = in.readDouble();
+                    lastTimestamp = in.readVLong();
+                }
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+        }
+
+        BytesRef serialize() {
+            ByteArrayOutputStream baos = new ByteArrayOutputStream();
+            OutputStreamStreamOutput out = new OutputStreamStreamOutput(baos);
+            try {
+                out.writeDouble(lastValue);
+                out.writeVLong(lastTimestamp);
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+            return new BytesRef(baos.toByteArray());
+        }
+
+        @Override
+        public void close() {}
+
+        void add(double v, long ts) {
+            if (Double.isNaN(lastValue) || ts > lastTimestamp) {
+                lastValue = v;
+                lastTimestamp = ts;
+            }
+        }
+
+        void add(SingleState other) {
+            if (Double.isNaN(lastValue)) {
+                if (Double.isNaN(other.lastValue) == false) {
+                    lastValue = other.lastValue;
+                    lastTimestamp = other.lastTimestamp;
+                }
+            } else if (Double.isNaN(other.lastValue) == false) {
+                if (lastTimestamp < other.lastTimestamp) {
+                    lastValue = other.lastValue;
+                    lastTimestamp = other.lastTimestamp;
+                } else {
+                    other.lastValue = lastValue;
+                    other.lastTimestamp = lastTimestamp;
+                }
+            }
+        }
+
+        void add(BytesRef other) {
+            add(new SingleState(other));
+        }
+
+        /** Extracts an intermediate view of the contents of this state.  */
+        @Override
+        public void toIntermediate(Block[] blocks, int offset) {
+            assert blocks.length >= offset + 1;
+            blocks[offset] = new ConstantBytesRefVector(serialize(), 1).asBlock();
+        }
+
+        Block evaluateLastValue(DriverContext driverContext) {
+            if (Double.isNaN(lastValue)) {
+                return Block.constantNullBlock(1);
+            }
+            return DoubleBlock.newConstantBlockWith(lastValue, 1);
+        }
+    }
+
+    static class GroupingState implements GroupingAggregatorState {
+        private ObjectArray<SingleState> states;
```

Comment:
I'd love to get this with two big arrays instead of ObjectArray. That just tends to use a bunch less memory. It does sort of scatter the values, but it's what we've been trying to do. It'd be lovely to have "struct shaped" big arrays, but that's not a thing at the moment.